plugin/session_buffer.py

```diff
@@ -376,5 +392,127 @@ def _present_diagnostics_async(
+
+    def _draw_semantic_tokens_async(self) -> None:
+        view = self.some_view()
+        if view is None:
+            return
+
+        semantic_tokens_mapping = SEMANTIC_TOKENS_MAP
+        # TODO this mapping better should only be updated when a plugin is registered/unregistered and then stored in the
+        #      Session, and not over and over again for each request
+        if self.session._plugin is not None:
+            semantic_tokens_mapping.update(self.session._plugin.semantic_tokens())
+
+        # TODO compare old and new regions to only remove regions that don't exist anymore
+        for sv in self.session_views:
+            for key in self.semantic_tokens_keys:
+                sv.view.erase_regions(key)
+        self.semantic_tokens.clear()
+        self.semantic_tokens_keys.clear()
+
+        scope_regions = dict()  # type: Dict[str, List[sublime.Region]]
+        prev_line = 0
+        prev_col = 0
+
+        for idx in range(0, len(self.semantic_tokens_data), 5):
+            delta_line = self.semantic_tokens_data[idx]
+            delta_start = self.semantic_tokens_data[idx + 1]
+            length = self.semantic_tokens_data[idx + 2]
+            token_type_encoded = self.semantic_tokens_data[idx + 3]
+            token_modifiers_encoded = self.semantic_tokens_data[idx + 4]
+            line = prev_line + delta_line
+            col = prev_col + delta_start if delta_line == 0 else delta_start
+            a = point_to_offset(Point(line, col), view)
+            b = a + length
+            r = sublime.Region(a, b)
+            prev_line = line
+            prev_col = col
+            token_type = self.semantic_token_types[token_type_encoded]  # type: ignore
+            token_modifiers = [self.semantic_token_modifiers[idx]  # type: ignore
+                               for idx, val in enumerate(reversed(bin(token_modifiers_encoded)[2:])) if val == "1"]
+            if token_type in semantic_tokens_mapping:
+                scope = semantic_tokens_mapping[token_type]
+                for entry in SEMANTIC_TOKENS_WITH_MODIFIERS_MAP:
+                    # TODO there must be a better way than to iterate through all entries
+                    if entry[0] == token_type and entry[1] in token_modifiers:
+                        scope = entry[2]
+                        break  # first match wins (in case of multiple modifiers)
+                scope += " meta.semantic-token.{}.lsp".format(token_type)
+            else:
+                # skip this token if no scope is defined for its token type
+                continue
+            self.semantic_tokens.append(SemanticToken(r, token_type, token_modifiers))
+            if scope in scope_regions:
+                scope_regions[scope].append(r)
+            else:
+                scope_regions[scope] = [r]
+
+        for scope, regions in scope_regions.items():
+            key = "lsp_{}".format(str(abs(hash(scope))))  # TODO find a better way to name region keys
```

Comment:
What's wrong with using `"lsp_{}".format(scope)` ?

Reply:
I wanted to avoid spaces and dots in the key, but actually that doesn't matter. Why do it simple if there is a complicated way ;) (sry, don't know if this expression makes sense in english)